### PR TITLE
Truncate long items in the SelectionList

### PR DESCRIPTION
**Pull request by abzokhattab**
### Fixed Issues

$ https://github.com/Expensify/App/issues/32384
PROPOSAL: https://github.com/Expensify/App/issues/32384#issuecomment-1836888387

### Tests

- [X] Verify that no errors appear in the JS console

1. Click on Profile -> workspaces -> Create a new workspace and modify the workspace name to be a long one
3. Click + > New chat > Room
4. Click Workspace
5. Verify that the long workspaces names are truncated and ending with `...`

Besides the previous test, All lists using SelectionList should truncate long items.
We can check a few of each type (single selection and multiple selection) to make sure both list items are functioning correctly:

Profile->timezone
Profile -> Pronouns
Workspace -> Members
Workspace -> Members -> Invite
Workspace -> Settings -> currency

### Offline tests

1. Click on Profile -> workspaces -> Create a new workspace and modify the workspace name to be a long one
3. Click + > New chat > Room
4. Click Workspace
5. Verify that the long workspaces names are truncated and ending with `...`

Besides the previous test, All lists using SelectionList should truncate long items.
We can check a few of each type (single selection and multiple selection) to make sure both list items are functioning correctly:

Profile->timezone
Profile -> Pronouns
Workspace -> Members
Workspace -> Members -> Invite
Workspace -> Settings -> currency

### QA Steps

- [x] Verify that no errors appear in the JS console

1. Click on Profile -> workspaces -> Create a new workspace and modify the workspace name to be a long one
3. Click + > New chat > Room
4. Click Workspace
5. Verify that the long workspaces names are truncated and ending with `...`

Besides the previous test, All lists using SelectionList should truncate long items.
We can check a few of each type (single selection and multiple selection) to make sure both list items are functioning correctly:

Profile->timezone
Profile -> Pronouns
Workspace -> Members
Workspace -> Members -> Invite
Workspace -> Settings -> currency

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
        - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

Screenshot 2023-12-05 at 11 33 14 PM

</details>

<details>
<summary>Android: mWeb Chrome</summary>

Screenshot 2023-12-06 at 12 16 44 AM

</details>

<details>
<summary>iOS: Native</summary>

Screenshot 2023-12-05 at 11 36 34 PM

</details>

<details>
<summary>iOS: mWeb Safari</summary>

Screenshot 2023-12-05 at 11 36 57 PM

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

image

</details>

<details>
<summary>MacOS: Desktop</summary>

image

</details>

### CODE REVIEW

**Comment by abzokhattab** [comment]
hi @s77rt, 
1. I added the `numberOfLines=1` in both the `item.text` and the `item.alternateText`
2. and please notice that I didn't add the tooltip component so let me know if we need to add it.

**Comment by s77rt** [comment]
@abzokhattab Please add the tooltip. It should be similar to that of UserListItem

**Comment by abzokhattab** [comment]
Okay done, the screenshots were updated.

File: src/components/SelectionList/RadioListItem.js [hunk 1]
```diff
@@ -1,17 +1,38 @@
...
     const styles = useThemeStyles();
     return (
         <View style={[styles.flex1, styles.alignItemsStart]}>
-            <Text style={[styles.optionDisplayName, isFocused ? styles.sidebarLinkActiveText : styles.sidebarLinkText, item.isSelected && styles.sidebarLinkTextBold]}>{item.text}</Text>
+            <Tooltip
+                shouldRender={showTooltip}
+                text={item.text}
+            >
+                <Text
+                    style={[styles.optionDisplayName, isFocused ? styles.sidebarLinkActiveText : styles.sidebarLinkText, item.isSelected && styles.sidebarLinkTextBold, styles.pre]}
```
Comment (s77rt) [inline src/components/SelectionList/RadioListItem.js:17]: The styles here and in `UserListItem` are mostly the same. Let's move those styles to `BaseListItem` and pass them as props: `textStyles` `alternateTextStyles`

**Review by s77rt — CHANGES_REQUESTED** [review]

**Comment by s77rt** [comment]
On Android the text does not take the full available space, but that bug is being worked on here https://github.com/Expensify/App/issues/21219 thus it's not a blocker

android

File: src/components/SelectionList/RadioListItem.js [hunk 2]
```diff
@@ -1,17 +1,39 @@
...
     const styles = useThemeStyles();
     return (
         <View style={[styles.flex1, styles.alignItemsStart]}>
-            <Text style={[styles.optionDisplayName, isFocused ? styles.sidebarLinkActiveText : styles.sidebarLinkText, item.isSelected && styles.sidebarLinkTextBold]}>{item.text}</Text>
+            <Tooltip
+                shouldRender={showTooltip}
+                text={item.text}
+            >
+                <Text
+                    style={StyleUtils.combineStyles(textStyle, item.isSelected && styles.sidebarLinkTextBold)}
```
Comment (s77rt) [inline src/components/SelectionList/RadioListItem.js:18]: The bold condition should be moved to the outer styles. Based on condition: `isUserItem || item.isSelected`

File: src/components/SelectionList/BaseListItem.js [hunk 3]
```diff
@@ -31,6 +31,20 @@ function BaseListItem({
     const isUserItem = lodashGet(item, 'icons.length', 0) > 0;
     const ListItem = isUserItem ? UserListItem : RadioListItem;
 
+    const textStyle = StyleUtils.combineStyles(
```
Comment (s77rt) [inline src/components/SelectionList/BaseListItem.js:34]: I don't think we need `StyleUtils.combineStyles`. Just return an array here

**Review by s77rt — CHANGES_REQUESTED** [review]

File: src/components/SelectionList/BaseListItem.js (hunk 3 above)
Comment (abzokhattab) [reply src/components/SelectionList/BaseListItem.js]: done

File: src/components/SelectionList/BaseListItem.js [hunk 4]
```diff
@@ -83,6 +83,14 @@ function BaseListItem({
                     )}
                     <ListItem
                         item={item}
+                        textStyle={[
+                            styles.optionDisplayName,
+                            isFocused ? styles.sidebarLinkActiveText : styles.sidebarLinkText,
+                            item.isSelected && styles.sidebarLinkTextBold,
```
Comment (s77rt) [inline src/components/SelectionList/BaseListItem.js:89]:
```suggestion
```
dupe

File: src/components/SelectionList/UserListItem.js [hunk 5]
```diff
@@ -7,7 +7,7 @@ import Tooltip from '@components/Tooltip';
 import useThemeStyles from '@styles/useThemeStyles';
 import {userListItemPropTypes} from './selectionListPropTypes';
 
-function UserListItem({item, isFocused = false, showTooltip}) {
+function UserListItem({item, textStyle, alternateTextStyle, showTooltip}) {
```
Comment (s77rt) [inline src/components/SelectionList/UserListItem.js:10]: Rename the props to  `textStyles`, `alternateTextStyles` and add them to `commonListItemPropTypes`

File: src/components/SelectionList/BaseListItem.js [hunk 6]
```diff
@@ -83,6 +83,14 @@ function BaseListItem({
...
+                            isUserItem || item.isSelected ? styles.sidebarLinkTextBold : null,
```
Comment (s77rt) [inline src/components/SelectionList/BaseListItem.js:90]:
```suggestion
                            (isUserItem || item.isSelected) && styles.sidebarLinkTextBold,
```
NAB

**Review by s77rt — CHANGES_REQUESTED** [review]

File: src/components/SelectionList/BaseListItem.js (hunk 3 above)
Comment (s77rt) [reply src/components/SelectionList/BaseListItem.js]: When we have a combination of arrays of objects and objects that needs to be flatten out. Here we only have objects.

File: src/components/SelectionList/BaseListItem.js (hunk 6 above)
Comment (abzokhattab) [reply src/components/SelectionList/BaseListItem.js]: Done

you suggested this change to be consistent with the existing code right? Or is there another reason for this change? just curious

Comment (abzokhattab) [reply src/components/SelectionList/BaseListItem.js:89]: Good catch ... reverted

**Comment by s77rt** [comment]
@abzokhattab Thanks!

File: src/components/SelectionList/BaseListItem.js [hunk 7]
```diff
@@ -83,7 +83,13 @@ function BaseListItem({
                     )}
                     <ListItem
                         item={item}
-                        isFocused={isFocused}
```
Comment (s77rt) [inline src/components/SelectionList/BaseListItem.js:86]: Can you please remove isFocused from commonListItemPropTypes

**Review by s77rt — CHANGES_REQUESTED** [review]
One last change

**Review by s77rt — APPROVED** [review]

**Review by iwiznia — APPROVED** [review]
The modified components are used in more places than the workspace list, right?
If so, can you add some tests to the description of the PR testing those other places please?
Then please ping me to merge this

**Comment by abzokhattab** [comment]
Okay @iwiznia I added some other tests.

File: src/components/SelectionList/BaseListItem.js [hunk 8]
```diff
@@ -83,7 +83,13 @@ function BaseListItem({
...
+                        textStyles={[
+                            styles.optionDisplayName,
+                            isFocused ? styles.sidebarLinkActiveText : styles.sidebarLinkText,
+                            isUserItem || item.isSelected ? styles.sidebarLinkTextBold : null,
+                            styles.pre,
+                        ]}
+                        alternateTextStyles={[styles.optionAlternateText, styles.textLabelSupporting, isFocused ? styles.sidebarLinkActiveText : styles.sidebarLinkText, styles.pre]}
```
Comment (s77rt) [inline src/components/SelectionList/BaseListItem.js:92]: The order of styles matter. We used to have `styles.textLabelSupporting` passed last so a font size of 13px is applied. But now with the new order another font size is being applied. This caused a regression (partly) https://github.com/Expensify/App/issues/32614

Comment (abzokhattab) [reply src/components/SelectionList/BaseListItem.js:92]: Oh, didn't notice it, should I create a PR to fix it?

Comment (iwiznia) [reply src/components/SelectionList/BaseListItem.js:92]: Yes please

Comment (abzokhattab) [reply src/components/SelectionList/BaseListItem.js:92]: Okay

Comment (s77rt) [reply src/components/SelectionList/BaseListItem.js:92]: This is fixed already

**Pull request merged**